### PR TITLE
feat: add compound border radius syntax support

### PR DESCRIPTION
**Pull request by ZijunYe**
### **PR Type**
Enhancement, Tests

### **Major Changed file**
- compoundBorderValue.js 
- compoundBorderValue.test.js 
- border.js 
- border.test.js 

### **Change Reason**
- Add support for assigning multiple `borderRadius` values (e.g., `sm` `md`  `lg` `xl`) in a single declaration.

- Allow a single `borderRadius` variable to define values for all four corners, replacing the need for separate values per direction.

### **Description**
- Added support for compound border radius syntax in the styled-system.

- Implemented `compoundBorderValue` transform for handling complex border radius values.

- Enhanced tests to validate compound border radius functionality.

- Introduced edge case handling for invalid or mixed border radius inputs.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>border.test.js</strong><dd><code>Add tests for compound border radius functionality</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/styled-system/src/config/__tests__/border.test.js

<li>Added tests for compound border radius syntax.<br> <li> Validated single, two, and four value scenarios.<br> <li> Tested mixed theme and raw value handling.

</details>

  </td>
  <td><a href="https://github.com/trendmicro-frontend/tonic-ui/pull/994/files#diff-4cb53327f1f1ce5a982e86df60e0614ee1e2ec9714196d67897d5bd0ee294d5d">+63/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>compoundBorderValue.test.js</strong><dd><code>Add tests for `compoundBorderValue` transform</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/styled-system/src/transforms/__tests__/compoundBorderValue.test.js

<li>Added unit tests for <code>compoundBorderValue</code> transform.<br> <li> Covered single, two, three, and four value scenarios.<br> <li> Included edge case tests for invalid inputs and whitespace.

</details>

  </td>
  <td><a href="https://github.com/trendmicro-frontend/tonic-ui/pull/994/files#diff-8acca1f088bd1024412b490294010b4af0d71fca5c240ac74a5908aa3d94e557">+71/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>border.js</strong><dd><code>Add compound border radius transform to configuration</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/styled-system/src/config/border.js

<li>Integrated <code>compoundThemeValueTransform</code> for border radius.<br> <li> Enhanced border radius configuration to support compound values.

</details>

  </td>
  <td><a href="https://github.com/trendmicro-frontend/tonic-ui/pull/994/files#diff-575f486c1767639ca5fab227edc4d50c16458aa523559dbc0595d90dd561df10">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>compoundBorderValue.js</strong><dd><code>Implement `compoundBorderValue` transform for border radius</code></dd></summary>
<hr>

packages/styled-system/src/transforms/compoundBorderValue.js

<li>Implemented <code>compoundBorderValue</code> transform for border radius.<br> <li> Added logic to normalize and process compound values.<br> <li> Handled edge cases like null, empty strings, and invalid inputs.

</details>

  </td>
  <td><a href="https://github.com/trendmicro-frontend/tonic-ui/pull/994/files#diff-709497f1013eb5f9f43185ed27f16aceb3e7c89a35872a3ffb82f3467f67c274">+96/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about Qodo Merge usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

### CODE REVIEW

File: packages/styled-system/src/transforms/compoundBorderValue.js [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+const hasOwnSafe = (obj, key) => {
+  if (isNullish(obj)) {
+    return false;
+  }
+
+  return Object.hasOwn
+    ? Object.hasOwn(obj, key)
+    : Object.prototype.hasOwnProperty.call(obj, key);
+};
+  /**
```
Comment (cheton) [inline packages/styled-system/src/transforms/compoundBorderValue.js:13]: redundant `/**` at line #13

File: packages/styled-system/src/transforms/compoundBorderValue.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    // Handle compound values
+    const parts = value.split(/\s+/).filter(Boolean);
+    const normalizedParts = normalizeRadiusValues(parts);
+
+    return normalizedParts
+      .map(part => {
+        // Handle raw numeric values with units
+        if (/^[0-9]+(\.[0-9]+)?(px|em|rem|%|vh|vw)$/.test(part)) {
```
Comment (cheton) [inline packages/styled-system/src/transforms/compoundBorderValue.js:77]:
```suggestion
        if (/^([0-9]*\.[0-9]+|[0-9]+)(cm|mm|in|px|pt|pc|em|ex|ch|rem|vw|vh|vmin|vmax|%)$/i.test(part))
```

Note: The regular expression above does not handle negative values. If we want to make a separate edge transform in additional to the corner transform, handling negative values will be an essential part.

1. The current regex doesn't match `.1rem`, you need to make the digits before the decimal optional.
    ```
    ([0-9]*\.[0-9]+|[0-9]+)
    ```
    * `([0-9]*\.[0-9]+)` allows .1, 0.1, 10.5, etc.
    * `[0-9]+` covers whole numbers like 1, 10, etc.

3. To ensure a complete coverage, it might be necessary to list all possible absolute and relative CSS units:
https://www.w3schools.com/cssref/css_units.php

Absolue units:
Unit | Description
:-- | :--
cm | centimeters
mm | millimeters
in | inches (1in = 96px = 2.54cm)
px | pixels (1px = 1/96th of 1in)
pt | points (1pt = 1/72 of 1in)
pc | picas (1pc = 12 pt)

Relative units:
Unit | Description
:-- | :--
em | Relative to the font-size of the element (2em means 2 times the size of the current font)
ex | Relative to the x-height of the current font (rarely used)
ch | Relative to the width of the "0" (zero)
rem | Relative to font-size of the root element
vw | Relative to 1% of the width of the viewport*
vh | Relative to 1% of the height of the viewport*
vmin | Relative to 1% of viewport's* smaller dimension
vmax | Relative to 1% of viewport's* larger dimension
% | Relative to the parent element

File: packages/styled-system/src/config/__tests__/border.test.js [hunk 3]
```diff
@@ -155,3 +155,66 @@ test('returns border styles', () => {
...
+    const style = border({
+      theme,
+      borderRadius: 'sm'
+    });
+    expect(style).toEqual({
+      borderTopLeftRadius: '0.125rem',
+      borderTopRightRadius: '0.125rem',
+      borderBottomLeftRadius: '0.125rem',
+      borderBottomRightRadius: '0.125rem'
+    });
```
Comment (cheton) [inline packages/styled-system/src/config/__tests__/border.test.js:179]: To improve readiness, it is recommended to maintain the following key order for consistency of `border-radius` declaration for all your tests:
```
borderTopLeftRadius
borderTopRightRadius
borderBottomRightRadius
borderBottomLeftRadius
```

File: packages/styled-system/src/config/border.js [hunk 4]
```diff
@@ -1,5 +1,6 @@
 import system from '../core/system';
 import positiveOrNegativeTransform from '../transforms/positiveOrNegative';
+import compoundThemeValueTransform from '../transforms/compoundBorderValue';
```
Comment (cheton) [inline packages/styled-system/src/config/border.js:3]: Consider renaming it to `compoundCornerTransform` because its is only intended to handle `border-radius` values in clockwise order.

For properties like `margin`, `padding`, `border-width`, `border-style`, `inset`, etc., we can implement a separate `compoundEdgeTransform` if we want to support them in the future.

**Comment by cheton** [comment]
Closing this PR as compound syntax support might not be fully supported in the current release, especially when considering various usage scenarios.
